packages/orm/src/EntityProxy.ts

```diff
@@ -0,0 +1,123 @@
+import { OneToManyCollection } from "./collections/OneToManyCollection";
+import { ManyToManyCollection } from "./collections/ManyToManyCollection";
+import { ManyToOneReference } from "./collections/ManyToOneReference";
+import { Entity } from "./EntityManager";
+
+// These proxies are passed into hooks during a flush to enable changes mid flush
+
+type Loader = OneToManyCollection<any, any> | ManyToManyCollection<any, any> | ManyToOneReference<any, any, any>;
+
+function isPlainFunction(obj: any) {
+  // This is absolutely gross, but there's seemingly no other way to figure out if a function is really a class
+  return typeof obj === "function" && !/^class\s/.test(obj.toString());
+}
+
+function proxyLoaderResult(result: any, flushSecret: number) {
+  if (Array.isArray(result)) {
+    return result.map((entity) => proxyEntity(entity, flushSecret));
+  } else if (typeof result === "object") {
+    return proxyEntity(result, flushSecret);
+  }
+
+  return result;
+}
+
+function proxyLoader(loader: Loader, entity: Entity, flushSecret: number) {
+  const prototype = Object.getPrototypeOf(loader);
+
+  return new Proxy(loader, {
+    get: (loader, property, receiver) => {
+      if (property === "flushSecret") {
+        return flushSecret;
+      }
+
+      if (property === "load") {
+        return function () {
+          return loader
+            .load({ beingDeleted: entity.__orm.deleted === "pending" })
+            .then((result) => proxyLoaderResult(result, flushSecret));
+        };
+      }
+
+      if (property === "get") {
+        return proxyLoaderResult(loader.get, flushSecret);
+      }
+
+      if (property === "entity") {
+        // this should be a proxied entity already
+        return entity;
+      }
+
+      let value: any;
+      const getter = prototype.__lookupGetter__(property);
+      if (getter !== undefined) {
+        value = getter.apply(receiver);
+      } else {
+        value = (loader as any)[property];
+      }
+
+      if (isPlainFunction(value)) {
+        return function (...args: any[]) {
+          return value.apply(receiver, ...args);
+        };
+      }
+
+      return value;
+    },
+  });
+}
+
+export function proxyEntity(entity: Entity, flushSecret: number): Entity {
+  const prototype = Object.getPrototypeOf(entity);
+  const loaderCache: any = {};
+
+  return new Proxy(entity, {
+    get(entity, property, receiver) {
+      if (property === "flushSecret") {
+        return flushSecret;
+      }
+
+      if (property === "proxyTarget") {
+        return entity;
+      }
+
+      let value: any;
+      const getter = prototype.__lookupGetter__(property);
+      if (getter !== undefined) {
+        value = getter.apply(receiver);
+      } else {
+        value = (entity as any)[property];
+      }
+
+      if (isPlainFunction(value)) {
+        return function (...args: any[]) {
+          return value.apply(receiver, ...args);
```

Comment:
I was refreshing my memory on proxies and was reading this:

https://exploringjs.com/es6/ch_proxies.html#sec_proxies-explained ... see the section "28.3.6.1 Wrapping an object affects this". They have an example that I flushed out to be:

```typescript
class Target {
  foo() {
    return this.bar();
  }
  bar() {
    return 2;
  }
}

describe("foo", () => {
  it("foo", () => {
    const target = new Target();
    const seenProps: string[] = [];
    const handler: ProxyHandler<any> = {
      get(target, prop) {
        seenProps.push(String(prop));
        return Reflect.get(target, prop);
      },
    };

    target.foo();
    expect(seenProps).toEqual([]);

    const proxy = new Proxy(target, handler);
    proxy.foo();
    expect(seenProps).toEqual(["foo", "bar"]);
  });
});
```

I.e. it looks like the `this` within proxied function calls is already the proxy... 

Does that simplify some of what you're trying to do?

Reply:
It seems crazy that this works:

```
class Target {
  ref = new Ref();
  foo() {
    return this.bar();
  }
  bar() {
    return this.ref;
  }
}
class Ref {
  zaz() {
    return (this as any).__secret;
  }
}

describe("foo", () => {
  it("foo", () => {
    const target = new Target();
    const seenProps: string[] = [];
    const handler: ProxyHandler<any> = {
      get(target, prop) {
        console.log(`GETGET ${String(prop)}`);
        seenProps.push(String(prop));
        if (prop === "__secret") {
          return (this as any).knownSecret;
        }
        const r = Reflect.get(target, prop);
        return typeof r === "object" && r ? new Proxy(r, handler) : r;
      },
      apply(target, prop) {
        console.log(`APPLY ${prop}`);
      }
    };

    target.foo();
    expect(seenProps).toEqual([]);

    (handler as any).knownSecret = "secret";
    const proxy = new Proxy(target, handler);
    const foundSecret = proxy.foo().zaz();
    expect(seenProps).toEqual(["foo", "bar", "ref", "zaz", "__secret"]);
    expect(foundSecret).toEqual("secret");
  });
});
```

It's like they built this aspect of proxies with our use case specifically in mind...